### DISCUSSION TITLE
SWO Encoding Limitations

### DISCUSSION THREAD

**Discussion by rijesha** [blackmagic-debug/blackmagic#1737] [General]
I was looking into the different supported SWO versions for different probes. 
It seems like some newer STM probes have NRZ support while older ones have only manchester support.

What is driving this? Isn't NRZ just standard UART, so couldn't any probe work with with it?

**Comment by dragonmux** [comment]
Unsure what you're defining as "older" vs "newer", but there are a couple of driving factors here. The first is that, as you noted, async mode needs a UART to really handle and there are several probes with pinouts that fail to put a UART RX input on that pin of their interface.

The other is that even if a newer design of probe does support this mode, if the firmware is for a lineup of them and with how the support is currently written (this is an item on our TODO list to fix), we cannot compile both kinds of support into a single firmware image yet, so the firmware on that linup is limited to the lowest common denominator.

**Reply by rijesha** [reply]
That makes sense. Yes my mistake on "newer" and "older" I was noticing that that BMP V23 is going to get NRZ support.

I am currently using a blackpill, but may switch to a blue pill to get the NRZ support. 

Somewhat related question, is there a noticeable change in debugging performance between BMP hardware? Is the latest BMP V23 or stlink v3 going to perform quicker than a blue pill?

**Reply by dragonmux** [reply]
BMP itself refers to the official hardware. In regards to debug performance, it's a STM32F103 based device so will have about the same throughput as anything else BMD supports based on that processor. However, what you get with buying BMP rather than, eg, a Blue Pill, is: a known quantity; variable voltage front-end (1.8V to 5V operation); unified debug connector; a ready to go product rather than a project. You'd also be supporting the project as BMP sales help pay for our maintaining the firmware.

ST-Link v3 probably does perform better than a Blue Pill because it's based on the STM32F723 which is just a more capable processor and also a much newer processor design. We're talking Cortex-M7 vs Cortex-M3, faster clocks, more Flash, more RAM, slightly better thought out bus structure, and a better USB device interface peripheral. NB though that any unit you can likely purchase will come with fairly new firmware which will have already set RDP2 (read protection level 2) so will not have a viable debug interface for the processor if anything goes wrong, and can only be reprogrammed via `stlink-tool`. The firmware is not well tested (yet) on these.
